File: frontend/src/schema.ts
import { gql } from "@apollo/client";

const COACH_FRAGMENT = gql`
  fragment Coach on CoachType {
    id
    phoneNumber
    firstName
    lastName
  }
`;

const STUDENT_FRAGMENT = gql`
  fragment Student on StudentType {
    id
    phoneNumber
    firstName
    lastName
  }
`;

const SLOT_FRAGMENT = gql`
  fragment Slot on SlotType {
    id
    startTime
    endTime
    coach {
      ...Coach
    }
    student {
      ...Student
    }
    call {
      id
    }
  }
  ${COACH_FRAGMENT}
  ${STUDENT_FRAGMENT}
`;

const CALL_FRAGMENT = gql`
  fragment Call on CallType {
    id
    slot {
      startTime
      endTime
    }
    coach {
      id
      firstName
      lastName
    }
    student {
      id
      firstName
      lastName
    }
    satisfactionScore
    notes
  }
`;

export const LIST_COACHES = gql`
  query ListCoaches {
    coaches {
      ...Coach
    }
  }
  ${COACH_FRAGMENT}
`;

export const LIST_STUDENTS = gql`
  query ListStudent {
    students {
      ...Student
    }
  }
  ${STUDENT_FRAGMENT}
`;

export const GET_SLOT = gql`
  query GetSlot($id: ID) {
    slot(id: $id) {
      ...Slot
    }
  }
  ${SLOT_FRAGMENT}
`;

export const GET_SLOT_COUNTS = gql`
  query GetSlotCounts($year: Int!, $month: Int, $coachId: ID, $studentId: ID) {
    slotCounts(
      year: $year
      month: $month
      coachId: $coachId
      studentId: $studentId
    )
  }
`;

export const LIST_SLOTS = gql`
  query ListSlots(
    $startTime: String
    $endTime: String
    $coachId: ID
    $studentId: ID
    $withCall: Boolean
  ) {
    slots(
      startTime: $startTime
      endTime: $endTime
      coachId: $coachId
      studentId: $studentId
      withCall: $withCall
    ) {
      ...Slot
    }
  }
  ${SLOT_FRAGMENT}
`;

export const CREATE_SLOT = gql`
  mutation CreateSlot($startTime: String!, $coachId: ID!) {
    createSlot(startTime: $startTime, coachId: $coachId) {
      slot {
        ...Slot
      }
    }
  }
  ${SLOT_FRAGMENT}
`;

export const BOOK_SLOT = gql`
  mutation BookSlot($id: ID!, $studentId: ID!) {
    bookSlot(id: $id, studentId: $studentId) {
      slot {
        ...Slot
      }
    }
  }
  ${SLOT_FRAGMENT}
`;

export const GET_CALL = gql`
  query GetCall($id: ID!) {
    call(id: $id) {
      ...Call
    }
  }
  ${CALL_FRAGMENT}
`;

export const CREATE_CALL = gql`
  mutation CreateCall($slotId: ID!, $satisfactionScore: Int!, $notes: String) {
    createCall(
      slotId: $slotId
      satisfactionScore: $satisfactionScore
      notes: $notes
    ) {
      call {
        ...Call
      }
    }
  }
  ${CALL_FRAGMENT}
`;

File: backend/scheduling/schema.py
import graphene
from graphene_django import DjangoObjectType
from scheduling.models import Slot, Call
from datetime import datetime
from dateutil.relativedelta import relativedelta
from django.db.models import Q


class SlotType(DjangoObjectType):
    class Meta:
        model = Slot


class CallType(DjangoObjectType):
    class Meta:
        model = Call


class Query(graphene.ObjectType):
    slot = graphene.Field(SlotType, id=graphene.ID())
    slots = graphene.List(
        SlotType,
        start_time=graphene.String(required=False),
        end_time=graphene.String(required=False),
        coach_id=graphene.ID(required=False),
        student_id=graphene.ID(required=False),
        with_call=graphene.Boolean(required=False),
    )
    slot_counts = graphene.JSONString(
        year=graphene.Int(),
        month=graphene.Int(),
        coach_id=graphene.ID(),
        student_id=graphene.ID(),
    )
    call = graphene.Field(CallType, id=graphene.ID())
    calls = graphene.List(CallType)

    def resolve_slot(root, info, id):
        return Slot.objects.filter(id=id).first()

    def resolve_slots(
        root,
        info,
        start_time=None,
        end_time=None,
        coach_id=None,
        student_id=None,
        with_call=None,
    ):
        qs = Slot.objects.all()
        if coach_id:
            qs = qs.filter(coach_id=coach_id)
        if student_id:
            qs = qs.filter(Q(student_id=student_id) | Q(student_id__isnull=True))
        if start_time:
            start_time = datetime.fromisoformat(start_time)
            qs = qs.filter(start_time__gte=start_time)
        if end_time:
            end_time = datetime.fromisoformat(end_time)
            qs = qs.filter(end_time__lt=end_time)
        if with_call is not None:
            qs = qs.filter(call__isnull=bool(with_call))
        return qs.order_by("start_time")

    def resolve_slot_counts(root, info, year, month, coach_id=None, student_id=None):
        qs = Slot.objects.all()
        if coach_id:
            qs = qs.filter(coach_id=coach_id)
        if student_id:
            qs = qs.filter(Q(student_id=student_id) | Q(student_id__isnull=True))

        start_date = datetime(year, month, 1)
        end_date = start_date + relativedelta(months=1)
        qs = qs.filter(start_time__gte=start_date, start_time__lt=end_date)

        slot_counts = {}
        for slot in qs:
            day_of_month = slot.start_time.day
            if day_of_month not in slot_counts:
                slot_counts[day_of_month] = {"booked": 0, "available": 0}
            if slot.student_id:
                slot_counts[day_of_month]["booked"] += 1
            else:
                slot_counts[day_of_month]["available"] += 1

        return slot_counts

    def resolve_call(root, info, id):
        return Call.objects.filter(id=id).first()

    def resolve_calls(root, info):
        return Call.objects.all()


class CreateSlotMutation(graphene.Mutation):
    class Arguments:
        start_time = graphene.String()
        coach_id = graphene.ID()

    slot = graphene.Field(SlotType)

    @classmethod
    def mutate(cls, root, info, start_time, coach_id):

        slot = Slot.objects.create(
            start_time=datetime.fromisoformat(start_time), coach_id=coach_id
        )
        return CreateSlotMutation(slot=slot)


class BookSlotMutation(graphene.Mutation):
    class Arguments:
        id = graphene.ID()
        student_id = graphene.ID()

    slot = graphene.Field(SlotType)

    @classmethod
    def mutate(cls, root, info, id, student_id):

        slot = Slot.objects.get(id=id)
        slot.student_id = student_id
        slot.save()
        return CreateSlotMutation(slot=slot)


class CreateCallMutation(graphene.Mutation):
    class Arguments:
        slot_id = graphene.ID()
        satisfaction_score = graphene.Int()
        notes = graphene.String()

    call = graphene.Field(CallType)

    @classmethod
    def mutate(cls, root, info, slot_id, satisfaction_score, notes=None):

        slot = Slot.objects.get(id=slot_id)
        call = Call.create_from_slot(
            slot, satisfaction_score=satisfaction_score, notes=notes
        )
        return CreateCallMutation(call=call)


class Mutation(graphene.ObjectType):
    create_slot = CreateSlotMutation.Field()
    book_slot = BookSlotMutation.Field()
    create_call = CreateCallMutation.Field()
